pkg/peertls/peertls_test.go

```diff
@@ -137,159 +120,536 @@ func TestVerifyPeerCertChains(t *testing.T) {
+		},
+		{
+			"existing key - newer timestamp",
+			newerExt,
+			nil,
+			nil,
+		},
+	}
+
+	for _, c := range cases {
+		t.Run(c.testID, func(t2 *testing.T) {
+			if !assert.NotNil(t, c.ext) {
+				t2.Fail()
+				t.FailNow()
+			}
+			err = revDB.Put(chain, c.ext)
+			if c.errClass != nil {
+				assert.True(t, c.errClass.Has(err))
+			}
+			if c.err != nil {
+				assert.Equal(t, c.err, err)
+			}
+
+			if c.err == nil && c.errClass == nil {
+				if !assert.NoError(t2, err) {
+					t2.Fail()
+					t.FailNow()
+				}
+				check(t2, c.ext)
+			}
+		})
+	}
+}
+
+func TestParseExtensions(t *testing.T) {
+	revokedLeafKeys, revokedLeafChain, err := newRevokedLeafChain()
 	assert.NoError(t, err)
+	_ = revokedLeafKeys
 
-	leafTemplate, err := LeafTemplate()
+	whitelistSignedKeys, whitelistSignedChain, err := newCertChain(3)
 	assert.NoError(t, err)
 
-	leafCert, err := NewCert(leafKey, rootKey, leafTemplate, caCert)
+	err = AddSignedCertExt(whitelistSignedKeys[0], whitelistSignedChain[0])
 	assert.NoError(t, err)
 
-	err = AddSignedLeafExt(rootKey, leafCert)
+	_, unrelatedChain, err := newCertChain(1)
 	assert.NoError(t, err)
 
-	whitelist := []*x509.Certificate{rootCert}
+	tmp, err := ioutil.TempDir("", os.TempDir())
+	if err != nil {
+		t.FailNow()
+	}
+
+	defer os.RemoveAll(tmp)
+	revDB, err := NewRevocationDBBolt(filepath.Join(tmp, "revocations.db"))
+	assert.NoError(t, err)
 
 	cases := []struct {
 		testID    string
 		config    TLSExtConfig
+		certChain []*x509.Certificate
 		whitelist []*x509.Certificate
-		expected  []result
+		errClass  *errs.Class
+		err       error
 	}{
 		{
-			"leaf whitelist signature",
+			"leaf whitelist signature - success",
 			TLSExtConfig{WhitelistSignedLeaf: true},
-			whitelist,
-			[]result{{true, nil}},
+			whitelistSignedChain,
+			[]*x509.Certificate{whitelistSignedChain[2]},
+			nil,
+			nil,
 		},
+		{
+			"leaf whitelist signature - failure (empty whitelist)",
+			TLSExtConfig{WhitelistSignedLeaf: true},
+			whitelistSignedChain,
+			nil,
+			&ErrVerifyCAWhitelist,
+			nil,
+		},
+		{
+			"leaf whitelist signature - failure",
+			TLSExtConfig{WhitelistSignedLeaf: true},
+			whitelistSignedChain,
+			unrelatedChain,
+			&ErrVerifyCAWhitelist,
+			nil,
+		},
+		{
+			"certificate revocation - single revocation ",
+			TLSExtConfig{Revocation: true},
+			revokedLeafChain,
+			nil,
+			nil,
+			nil,
+		},
+		{
+			"certificate revocation - multiple, serial revocations",
+			TLSExtConfig{Revocation: true},
+			func() []*x509.Certificate {
+				rev := new(Revocation)
+				err = rev.Unmarshal(revokedLeafChain[0].ExtraExtensions[0].Value)
+				assert.NoError(t, err)
+
+				time.Sleep(1 * time.Second)
+				_, chain, err := revokeLeaf(revokedLeafKeys, revokedLeafChain)
+				assert.NoError(t, err)
+
+				err = rev.Unmarshal(chain[0].ExtraExtensions[0].Value)
+				assert.NoError(t, err)
+
+				return chain
+			}(),
+			nil,
+			nil,
+			nil,
+		},
+		// TODO: test multiple simultaneous extensions
```

Comment:
what are simultaneous extensions?

Reply:
By that i just mean we probably want to test that multiple extension handlers work together as expected.